### DISCUSSION TITLE
Adding custom keymap.c code for key overrides

### DISCUSSION THREAD

**Discussion by laurapigeon** [manna-harbour/miryoku#279] [Q&A] [answered]
I'm using [this from the qmk guide](https://docs.qmk.fm/#/feature_key_overrides?id=key-overrides) to try to add the simple delete override example to my miryoku layout [in this repo](https://github.com/laurapigeon/miryoku_qmk_sturdy/tree/miryoku). all of my layout changes so far have been in [users -> custom_config.h](https://github.com/laurapigeon/miryoku_qmk_sturdy/blob/miryoku/users/manna-harbour_miryoku/custom_config.h). based on [the customisation documentation](https://github.com/manna-harbour/miryoku/discussions/85), ive copied across the files from layouts -> config.h and keymap.c to [keyboards -> keymaps](https://github.com/laurapigeon/miryoku_qmk_sturdy/tree/miryoku/keyboards/dasbob/keymaps/manna-harbour_miryoku). i added `KEY_OVERRIDE_ENABLE = yes` to the keyboard rules.mk file (and by default its already in the users one), and added the example code to the new keyboard keymap.c file. this last step causes errors when compiling:
```./keyboards/dasbob/keymaps/manna-harbour_miryoku/keymap.c:6:7: error: unknown type name 'key_override_t'
    6 | const key_override_t delete_key_override = ko_make_basic(MOD_MASK_SHIFT, KC_BSPC, KC_DEL);
      |       ^~~~~~~~~~~~~~
./keyboards/dasbob/keymaps/manna-harbour_miryoku/keymap.c:6:44: error: implicit declaration of function 'ko_make_basic' [-Werror=implicit-function-declaration]
    6 | const key_override_t delete_key_override = ko_make_basic(MOD_MASK_SHIFT, KC_BSPC, KC_DEL);
      |                                            ^~~~~~~~~~~~~
./keyboards/dasbob/keymaps/manna-harbour_miryoku/keymap.c:6:58: error: 'MOD_MASK_SHIFT' undeclared here (not in a function)
    6 | const key_override_t delete_key_override = ko_make_basic(MOD_MASK_SHIFT, KC_BSPC, KC_DEL);
      |                                                          ^~~~~~~~~~~~~~
./keyboards/dasbob/keymaps/manna-harbour_miryoku/keymap.c:6:74: error: 'KC_BSPC' undeclared here (not in a function)
    6 | const key_override_t delete_key_override = ko_make_basic(MOD_MASK_SHIFT, KC_BSPC, KC_DEL);
      |                                                                          ^~~~~~~
./keyboards/dasbob/keymaps/manna-harbour_miryoku/keymap.c:6:83: error: 'KC_DEL' undeclared here (not in a function)
    6 | const key_override_t delete_key_override = ko_make_basic(MOD_MASK_SHIFT, KC_BSPC, KC_DEL);
      |                                                                                   ^~~~~~
./keyboards/dasbob/keymaps/manna-harbour_miryoku/keymap.c:9:7: error: unknown type name 'key_override_t'
    9 | const key_override_t **key_overrides = (const key_override_t *[]){
      |       ^~~~~~~~~~~~~~
./keyboards/dasbob/keymaps/manna-harbour_miryoku/keymap.c:9:47: error: unknown type name 'key_override_t'
    9 | const key_override_t **key_overrides = (const key_override_t *[]){
      |                                               ^~~~~~~~~~~~~~
In file included from quantum/quantum.h:118,
                 from ./.build/obj_dasbob/src/default_keyboard.h:27,
                 from users/manna-harbour_miryoku/manna-harbour_miryoku.c:6,
                 from quantum/keymap_introspection.c:9:
quantum/process_keycode/process_key_override.h:92:31: error: conflicting types for 'key_overrides'; have 'const key_override_t **'
   92 | extern const key_override_t **key_overrides;
      |                               ^~~~~~~~~~~~~
./keyboards/dasbob/keymaps/manna-harbour_miryoku/keymap.c:9:24: note: previous definition of 'key_overrides' with type 'const int **'
    9 | const key_override_t **key_overrides = (const key_override_t *[]){
      |                        ^~~~~~~~~~~~~
users/manna-harbour_miryoku/manna-harbour_miryoku.c:56:24: error: conflicting types for 'key_overrides'; have 'const key_override_t **'
   56 | const key_override_t **key_overrides = (const key_override_t *[]){
      |                        ^~~~~~~~~~~~~
./keyboards/dasbob/keymaps/manna-harbour_miryoku/keymap.c:9:24: note: previous definition of 'key_overrides' with type 'const int **'
    9 | const key_override_t **key_overrides = (const key_override_t *[]){
      |                        ^~~~~~~~~~~~~
```
did i place the rule in the wrong file? or the example code? thanks

**Comment by laurapigeon** [comment]
I think the issue I was having here is related to mixing keyboard keymap.c and user keymap.c (manna-harbour_miryoku.c). Since adding my overrides to the latter I've managed to get at least one override to work.

**Comment by manna-harbour** [comment, marked as answer]
The customisation docs recommend using the keyboard keymap.c for custom hardware support. Customising the keymap outside of what's supported by custom_config.h isn't described yet.

The plan is to add a custom_keymap.c for that purpose. However, certain keymap features may only work when in the same file as the actual keymap, so that may not be a general solution.

For now, arbitrary keymap customisations will need be made in manna-harbour_miryoku.c and manna-harbour_miryoku.h.

For key overrides in particular, as they are [already in use](https://github.com/manna-harbour/miryoku_qmk/blob/df1690ad257e7b0d1bf9450258f7baebfee7be9a/users/manna-harbour_miryoku/manna-harbour_miryoku.c#L56) they will need to be incorporated there. In future, support could be added so additions can be declared in custom_config.h.

**Reply by manna-harbour** [reply]
Added to https://github.com/manna-harbour/miryoku/issues/94
